### ISSUE TITLE
EF Core 3 and Data Annotation

### ISSUE THREAD

**Issue by clarghi** [ReversePOCO/EntityFramework-Reverse-POCO-Code-First-Generator#555]
Hi,
I use this utility from a long time. Yesterday I have buyed the licence to use with EF Core 3 in a .NETSTANDARD 2.0.3 project.
It works fine, but all data attributes are lost, and the settings UseDataAnnotations{Schema] = true; seems not to be available yet.
Why?
The namespace System.ComponentModel.DataAnnotations; is usable in .NET STANDARD (I have tried with a xamarin android project and worked well). Without this I cannot convert the old EF6 projects to NetStandard because I use Data Annotation for validation in all my projects.

B.R.
Cristiano Larghi

**Comment by sjh37** [comment]
Hi @clarghi 

The Data annotations are not as complete or as comprehensive as the fluent configuration so there is no real need to keep them. The `[Display]` name attribute is handy for forms and can be added back in easily (see below). You can also manually add data annotations within the `Settings.UpdateColumn` callback. For example:

```c#
// Adding custom data annotations (include your extra
// namespaces in Settings.AdditionalNamespaces)
column.Attributes.Add($"[Display(Name = \"{column.DisplayName}\")]");
if (column.IsPrimaryKey)
    column.Attributes.Add("[Key]");

// Adding length data annotations
if (!column.IsMaxLength && column.MaxLength > 0)
{
    column.Attributes.Add(string.Format("[MaxLength({0})]", column.MaxLength));

    if (column.PropertyType.Equals("string", StringComparison.InvariantCultureIgnoreCase))
        column.Attributes.Add(string.Format("[StringLength({0})]", column.MaxLength));
}

if (column.IsMaxLength) 
    column.Attributes.Add("[MaxLength]");
```

The fluent configuration has everything you need to have a correctly configured EF ORM model.
Are there any data annotations you'd specifically like to see?

**Comment by clarghi** [comment]
Hi,
An example to clarify.
With v2:

[DatabaseGenerated(DatabaseGeneratedOption.None)]
        [Column(@"PK_Area", Order = 1, TypeName = "uniqueidentifier")]
        [Index(@"PK_Areas", 1, IsUnique = true, IsClustered = true)]
        [Required]
        [Key]
        public System.Guid PK_Area { get; set; } // PK_Area (Primary key)

With v3 (this is EF6 because the problem is the same):

public Guid PkArea { get; set; } // PK_Area (Primary key)

In this example I use RequiredEx for both client and server-side validation (but I use almost validation provided as MaxLength or IsNumeric and some customs created by me. In this latest case I add them manually in a derived class).

I cannot add manually every attribute in every table because there are some hundreds of tables, views and SP.

I have seen that now these properties are attached from the [Tbl]Configuration file. I must test if, using this logic, the validations works correctly.

N.b. I use my version of the Display (so this is not a problem, I added this in the derived classes).

**Comment by sjh37** [comment]
Adding the following into `Settings.UpdateColumn` should be most of what you need:

```c#
column.Attributes.Add($"[Display(Name = \"{column.DisplayName}\")]");
if (column.IsPrimaryKey)
    column.Attributes.Add("[Key]");

// Adding length data annotations
if (!column.IsMaxLength && column.MaxLength > 0)
{
    column.Attributes.Add(string.Format("[MaxLength({0})]", column.MaxLength));

    if (column.PropertyType.Equals("string", StringComparison.InvariantCultureIgnoreCase))
        column.Attributes.Add(string.Format("[StringLength({0})]", column.MaxLength));
}

// Adding required data annotation
if (!column.IsNullable && !column.IsComputed)
{
    if (column.PropertyType.Equals("string", StringComparison.InvariantCultureIgnoreCase) && column.AllowEmptyStrings)
        column.Attributes.Add("[Required(AllowEmptyStrings = true)]");
    else
        column.Attributes.Add("[Required]");
}

if (column.IsMaxLength) 
    column.Attributes.Add("[MaxLength]");

if (column.IsRowVersion)
    column.Attributes.Add("[Timestamp]");
```

Also set the following at line 91:
```c#
Settings.AdditionalNamespaces = new []{ "System.ComponentModel.DataAnnotations" };
```

**Comment by clarghi** [comment]
Seems perfect, thanks!!!

**Comment by sjh37** [comment]
Cool. I've added those as examples to the database.tt so others can enable it too in future.

**Comment by clarghi** [comment]
Hi,
I use this issue because is related to the update from 2.x to 3.x. and Data Annotations.
With 3.1.0 version I cannot set a GUID with the attribute [DatabaseGenerated(DatabaseGeneratedOption.Identity)] because is overridden in the Configuration file created.
I use MetadataType attribute and it worked with 2.x version.
E.g. given a table with a GUID PK_Sys_Config with newID() setted as default value in the DB table I create the table with this generator as partial, then I have:

`[MetadataType(typeof(GLB_Sys_Config_Metadata))]
    public partial class GLB_Sys_Config {
    }
    public class GLB_Sys_Config_Metadata {
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public Guid PK_Sys_Config {
            get; set;
        }
    }`
This work in 2.x but doesn't work anymore in 3.x (and so all my GUID PK have now all zeroe's and crash with the second insert because are uniques).
The problem is generated by the autogenerated file GLB_Sys_ConfigConfiguration because in the constructor now there is:

`Property(x => x.PK_Sys_Config).HasColumnName(@"PK_Sys_Config").HasColumnType("uniqueidentifier").IsRequired().HasDatabaseGeneratedOption(DatabaseGeneratedOption.None);`

And this overwrite the previous setting.
Removing the last method call (.HasDatabaseGeneratedOption(DatabaseGeneratedOption.None)) resolve the issue and my code returns to work correctly.

How can I resolve this problem? Or there is an alternative (obviously without modify manually the self-created POCO classes) to use the MetadataType Attribute?

An alternative could be to create the DB fields with its default values: if the property is generated with this (newID() in this context) the problem will be resolved.

Thanks in advance
B.r.
Cristiano Larghi

**Comment by clarghi** [comment]
Ok, I have worked in a wrong way: after three years I use this utility I have found that support the Default values. I have never seen these settings because the user I use to generate doesn't generate the Defaults (while the SA, tested now, do it), I think for a missing permission :-/

So, for me the problem is resolved (I remove all Metadatatypes), but I think that will be better if MetaDataTypes can be used because is a "standard" way to add attributes to a property.

B.r.
Cristiano Larghi
